src/utils/question.rs

```diff
@@ -121,10 +121,13 @@ impl<'a> Confirmation<'a> {
             (Some(placeholder), Some(subs)) => Cow::Owned(self.prompt.replace(placeholder, subs)),
         };
 
+        let _locks = lock_and_flush_output_stdio()?;
+        let mut stdin_lock = stdin().lock();
```

Comment:
> then since you can ask for 1 input at a time, stdin would also be locked and block other threads that wanted to ask for user input since they can't do any work anyway.

IIRC (might not) this wasn't blocking stderr so logs could still appear in front of the question, even with stdin locked. Now the `ask` method locks all output, which makes sense because we shouldn't let any other thread output something in-between this STDIN interaction.

You may notice that I lock stdout and stderr multiple times in some of these execution paths, that was intentional to make sure that the user warnings would stay close to the question, and that's possible because StderrLock and StdoutLock both use ReentrantLock instead of a regular Mutex.

---

On second thought, the piece of code above almost introduces a rare deadlock bug, if stdin was locked before stderr.

That would happen if:
1. Thread 1 locked stderr and then requested stdin lock
2. Thread 2 locked stdin and then requested stderr lock

1. is possible, and if I switch the order of these two lines, 2. would also be possible, simultaneously (1. still possible in some utils::question functions).

This all seems a bit error-prone so before merging I'll re-review my code to make sure deadlocks aren't possible, maybe I can do that by Sunday or next week.

Reply:
Yeah it was blocking both initially, its the snippet I linked (should be from the commit just before you started pushing)

I would've never caught that not going to lie. This is something we could maybe solve with [fences](https://doc.rust-lang.org/std/sync/atomic/fn.fence.html)? I have to read how those work again, not sure if this is possible.

~Another thing that might work but is significantly more weird would be to have a static atomic bool and a [CAS op](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html#method.compare_exchange) to indicate to other threads that you have exclusive access to both stdout and stdin. That would make both locks be indicated as one atomic operation making the deadlock impossible. Each thread would then need to try and change the booleans value in a loop.~ It is at this point that I realized I am describing implementing a Mutex which would probably be easier.

Reply:
Just to clarify, for the 2nd point I was thinking of something like this:

```rs
static LOCKED: AtomicBool = AtomicBool::new(false);

//...
    pub fn ask(&self, substitute: Option<&'a str>) -> crate::Result<bool> {
        // Set `LOCKED` from false to true, acquiring the lock
        loop {
            let prev = LOCKED.compare_exchange(false, true, Ordering::AcqRel, Ordering::Relaxed);

            if let Ok(false) = prev {
                break;
            }
        }

        // Do stuff with stdin and out/err here

        // Set `LOCKED` to false, releasing the lock before returning from this method
        LOCKED.store(false, Ordering::Release);
    }
```